api/Clspv.cpp

```diff
@@ -0,0 +1,144 @@
+  }
+  *container = new (std::nothrow) ClspvContainerImpl;
+  if (!(*container)) {
+    return CLSPV_OUT_OF_HOST_MEM;
+  }
+  return CLSPV_SUCCESS;
+}
+
+void clspvDestroyContainer(ClspvContainer *container) {
+  std::scoped_lock<std::mutex> sl(g_clspvMutex);
+  if (container == nullptr || *container == nullptr) {
+    return;
+  }
+  delete *container;
+  *container = nullptr;
+}
+
+ClspvError clspvCompileFromSourcesString(
+    ClspvContainer *container, const size_t program_count,
+    const size_t *program_sizes, const char **programs, const char *options,
+    char **output_binary, size_t *output_binary_size, const char **output_log) {
+  if (container == nullptr || *container == nullptr) {
+    return CLSPV_INVALID_CONTAINER;
+  }
+  if (programs == nullptr || program_count == 0 || output_binary == nullptr ||
+      output_binary_size == nullptr) {
+    return CLSPV_INVALID_ARG;
+  }
+
+  int err = CLSPV_SUCCESS;
+  std::string sOptions(options ? options : "");
+  std::vector<std::string> vPrograms(program_count);
+  for (size_t i = 0; i < program_count; ++i) {
+    if (programs[i] == nullptr) {
+      return CLSPV_ERROR;
+    }
+    if (program_sizes && program_sizes[i] != 0) {
+      vPrograms[i].assign(programs[i], program_sizes[i]);
+    } else {
+      vPrograms[i].assign(programs[i]);
+    }
+  }
+
+  {
+    std::scoped_lock<std::mutex> sl((*container)->containerMutex);
+
+    if (!(*container)->buildLog.empty()) {
+      (*container)->buildLog.clear();
+    }
+    if (!(*container)->binary.empty()) {
+      (*container)->binary.clear();
+        }
+
+    err = clspv::CompileFromSourcesString(
+        vPrograms, sOptions, &(*container)->binary, &(*container)->buildLog);
+    if (output_log != nullptr) {
+      *output_log = (*container)->buildLog.c_str();
+    }
+
+    if (err != 0) {
+        return CLSPV_ERROR;
+    }
+
+    // Return binary as a char*
+    *output_binary = reinterpret_cast<char*>((*container)->binary.data());
+    *output_binary_size = (*container)->binary.size() * sizeof(uint32_t);
+  }
+
+  return CLSPV_SUCCESS;
+}
+
+ClspvError clspvParseReflectionData(ClspvContainer *container,
+                                    const char *spv_env_str,
+                                    uint32_t *spv_binary, size_t spv_words,
+                                    const char **output_data) {
+  if (container == nullptr || *container == nullptr) {
+    return CLSPV_INVALID_CONTAINER;
+  }
+  if (output_data == nullptr || spv_words == 0 || spv_binary == nullptr ||
+      spv_env_str == nullptr) {
+    return CLSPV_INVALID_ARG;
+  }
+
+  spv_target_env spvEnv;
+  if (!spvParseTargetEnv(spv_env_str, &spvEnv)) {
+    return CLSPV_ERROR;
+  }
+
+  {
+    std::scoped_lock<std::mutex> sl((*container)->containerMutex);
+    if ((*container)->reflectionData.tellp() != std::streampos(0)) {
+      (*container)->reflectionData.clear();
+    }
+
+    if (!clspv::ParseReflection(spv_binary, spv_words, spvEnv,
```

Comment:
This is not the right function to use. It has intentionally not been kept up-to-date with interface changes. Users of clspv are responsible for parsing the embedded reflection data in the SPIR-V.

Reply:
Ack.